picard/browser/browser.py

```diff
@@ -38,18 +39,52 @@
 from picard.util import mbid_validate
 
 
-def response(code):
+def response(code, origin=None):
     if code == 200:
         resp = '200 OK'
     elif code == 400:
         resp = '400 Bad Request'
     else:
         resp = '500 Internal Server Error'
+    if origin:
+        allowOrigin = 'Access-Control-Allow-Origin: {}\r\n'.format(origin)
+    else:
+        allowOrigin = ''
     return bytearray(
         'HTTP/1.1 {}\r\n'
         'Cache-Control: max-age=0\r\n'
+        '{}'
         '\r\n'
-        'Nothing to see here.\r\n'.format(resp), 'ascii')
+        'Nothing to see here.\r\n'.format(resp, allowOrigin), 'ascii')
+
+
+def _read_origin(conn):
+    while True:
+        line = conn.readLine().data()
+        if not line or line == b'\r\n':
+            break
+        if line.startswith(b'Origin:'):
+            header, origin = line.decode().split()
+            log.debug('Browser sent origin %s', origin)
+            if _is_valid_origin(origin):
+                return origin
+            else:
+                return None
+    return None
+
+
+RE_VALID_ORIGINS = re.compile(r'(.*\.)?musicbrainz.org')
```

Comment:
```suggestion
RE_VALID_ORIGINS = re.compile(r'^(?:[^\.]+\.)*musicbrainz\.org$')
```

This regex is incorrect (when used with `re.match()`):

```
>>> RE_VALID_ORIGINS = re.compile(r'(.*\.)?musicbrainz.org')
>>> RE_VALID_ORIGINS.match('a.b.musicbrainzzorg')
<_sre.SRE_Match object; span=(0, 19), match='a.b.musicbrainzzorg'>
>>> RE_VALID_ORIGINS.match('a.b.musicbrainz.org')
<_sre.SRE_Match object; span=(0, 19), match='a.b.musicbrainz.org'>
>>> RE_VALID_ORIGINS.match('not.musicbrainzborg.com')
<_sre.SRE_Match object; span=(0, 19), match='not.musicbrainzborg'>
>>> RE_VALID_ORIGINS.match('....musicbrainz.org')
<_sre.SRE_Match object; span=(0, 19), match='....musicbrainz.org'>
>>> RE_VALID_ORIGINS.match('musicbrainz.org')
<_sre.SRE_Match object; span=(0, 15), match='musicbrainz.org'>
```

A safer regexp would be: 

```
>>> RE_VALID_ORIGINS = re.compile(r'^(?:[^\.]+\.)*musicbrainz\.org$')
>>> RE_VALID_ORIGINS.match('a.b.musicbrainzzorg')
>>> RE_VALID_ORIGINS.match('a.b.musicbrainz.org')
<_sre.SRE_Match object; span=(0, 19), match='a.b.musicbrainz.org'>
>>> RE_VALID_ORIGINS.match('not.musicbrainzborg.com')
>>> RE_VALID_ORIGINS.match('....musicbrainz.org')
>>> RE_VALID_ORIGINS.match('musicbrainz.org')
<_sre.SRE_Match object; span=(0, 15), match='musicbrainz.org'>
```

It allows subdomains, escape last dot, matches start & end of the string, disallows string with dot at start or multiple dots. Also there's no need to capture first part (`(reg) -> (?:reg)`)